lib/Doctrine/DBAL/Connection.php

```diff
@@ -750,6 +750,66 @@ public function update($tableExpression, array $data, array $identifier, array $
         return $this->executeUpdate($sql, $values, $types);
     }
 
+    /**
+     * Executes an SQL UPDATE statement on a row, or an insert if it doesn't exist.
+     *
+     * Table expression and columns are not escaped and are not safe for user-input.
+     *
+     * @param string $tableExpression  The expression of the table to update quoted or unquoted.
+     * @param array  $data       An associative array containing column-value pairs.
+     * @param array  $identifier The update criteria. An associative array containing column-value pairs.
+     * @param array  $types      Types of the merged $data and $identifier arrays in that order.
+     *
+     * @return integer The number of affected rows.
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function upsert($tableExpression, array $data, array $identifier, array $types = array())
+    {
+        $qb = $this->createQueryBuilder()
+            ->select($this->getDatabasePlatform()->getCountExpression('*'))
```

Comment:
You would need the count if you wanted to prevent `upsert` accidentally updating multiple rows here

Reply:
It's a slippery slope. If the `UPSERT` query is not specific enough so that it can match more than one record, then it's invalid for this use case, no matter if zero, one or more matching records exist.

Could you give an example of a table and/or query where `INSERT` would succeed but `UPDATE` would affect multiple rows?

Reply:
That's also a good point